### PR TITLE
'Division' field name and picklist items missing

### LINKED ISSUE #735: 'Division' field name and picklist items missing
The division field caption and information in accession form doesn't show up
Screenshot (134)

### PR DESCRIPTION
**Pull request by maxxxxxdlp**
Fixes #735

### CODE REVIEW

**Review by grantfitzsimmons — CHANGES_REQUESTED** [review]
accessionlist
This is the 2021_11_25_20_18 database on the test panel. The accession form still is missing the division name. Tested with Specify7 branch issue-735.

**Review by grantfitzsimmons — CHANGES_REQUESTED** [review]
Bad news. With this branch, you cannot use the Data Entry feature. It does not respond
badbad
I tested this with freshfish and 2021_11_25_20.

On the bright side, the picklist items and Division field name are present.
accession

File: specifyweb/frontend/js_src/lib/divisionfieldcbx.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            return this.divisions;
+        },
+        render: function() {
+            if(typeof this.divisionsPromise === 'object'){
+                this.divisionsPromise.then((divisions)=>{
+                    this.divisions = divisions.map((division)=>({
+                        value: division.get('id'),
+                        title: division.get('name')
+                    }));
+                    this.divisionsPromise = undefined;
```
Comment (benanhalt) [inline specifyweb/frontend/js_src/lib/divisionfieldcbx.js]: Don't do this! One of the main features of promises is that invoking `then` on an already resolved promise will invoke the continuation immediately so you don't have to keep track of the state manually!

File: specifyweb/frontend/js_src/lib/divisionfieldcbx.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                    this.render();
+                });
+                return this;
+            }
+
+            var options = this.getPickListFields().map(function(item) {
+                return $('<option>').attr('value', item.value).text(item.title)[0];
+            });
+            this.$el.empty().append(options).prop('disabled', options.length < 1);
+            this.$el.val(this.resource.get('fieldname'));
```
Comment (benanhalt) [inline specifyweb/frontend/js_src/lib/divisionfieldcbx.js]: `'fieldname'`? How is that supposed to work?

**Comment by benanhalt** [comment]
@grantfitzsimmons, would you please test this with a database that has more than one division? Make sure that the division can be set, saved, reloaded with the correct value appearing, and then changed to a new value, saved and reloaded.

File: specifyweb/frontend/js_src/lib/divisionfieldcbx.js (hunk 1 above)
Comment (maxxxxxdlp) [reply specifyweb/frontend/js_src/lib/divisionfieldcbx.js]: Oh my god, why didn't I know about this sooner!
I always though that this code is invalid:
```js
const a = Promise.resolve();
a.then(console.log);
a.then(console.log);  // Error?
```

Comment (benanhalt) [reply specifyweb/frontend/js_src/lib/divisionfieldcbx.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#:~:text=A%20promise%20can%20participate%20in%20more%20than%20one%20nesting

**Review by grantfitzsimmons — CHANGES_REQUESTED** [review]
**The data entry forms no longer work on this branch!**

After testing this with multiple divisions, the division can be set, saved, reloaded multiple times and preserve its original value. This issue appears to be solved.

**Comment by benanhalt** [comment]
The data entry forms not working is a separate issue that has been fixed on another branch.

**Pull request merged**